blockchain/chain_sync/src/sync.rs

```diff
@@ -0,0 +1,493 @@
+use address::Address;
+use amt::AMT;
+use async_std::prelude::*;
+use async_std::stream::Stream;
+use async_std::sync::{Receiver, Sender};
+use blocks::{Block, BlockHeader, FullTipset, TipSetKeys, Tipset, TxMeta};
+use chain::ChainStore;
+use cid::Cid;
+use crypto::is_valid_signature;
+use db::Error as DBError;
+use encoding::{Cbor, Error as EncodingError};
+use forest_libp2p::{NetworkEvent, NetworkMessage};
+use futures::{select, FutureExt};
+use ipld_blockstore::BlockStore;
+use libp2p::core::PeerId;
+use log::info;
+use lru::LruCache;
+use message::Message;
+use num_bigint::BigUint;
+use pin_project::pin_project;
+use state_manager::StateManager;
+use state_tree::{HamtStateTree, StateTree};
+use std::cmp::min;
+use std::collections::HashMap;
+use std::future::Future;
+use std::sync::Arc;
+use std::{
+    pin::Pin,
+    task::{Context, Poll},
+};
+
+#[derive(PartialEq, Debug, Clone)]
+/// Current state of the ChainSyncer
+enum SyncState {
+    /// No useful peers, bootstrapping network to be able to make BlockSync requests
+    Stalled,
+
+    /// Syncing to checkpoint (using BlockSync for now)
+    _SyncCheckpoint,
+
+    /// Receive new blocks from the network and sync toward heaviest tipset
+    _ChainCatchup,
+
+    /// Once all blocks are validated to the heaviest chain, follow network
+    /// by receiving blocks over the network and validating them
+    _Follow,
+}
+
+#[pin_project]
+pub struct ChainSyncer<'db, DB, ST> {
+    /// Syncing state of chain sync
+    _state: SyncState,
+
+    /// manages retrieving and updates state objects
+    state_manager: StateManager<'db, DB, ST>,
+
+    /// manages sync buckets
+    sync_manager: SyncManager,
+
+    /// access and store tipsets / blocks / messages
+    chain_store: ChainStore<'db, DB>,
+
+    /// Context to be able to send requests to p2p network
+    network: SyncNetworkContext,
+
+    /// the known genesis tipset
+    _genesis: Tipset,
+
+    /// Bad blocks cache, updates based on invalid state transitions.
+    /// Will mark any invalid blocks and all childen as bad in this bounded cache
+    bad_blocks: LruCache<Cid, String>,
+
+    /// Channel for incoming network events to be handled by syncer
+    #[pin]
+    network_rx: Receiver<NetworkEvent>,
+}
+
+// TODO probably remove this in the future, polling as such probably unnecessary
+impl<'db, DB, ST> Future for ChainSyncer<'db, DB, ST> {
+    type Output = ();
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        match self.project().network_rx.poll_next(cx) {
+            Poll::Ready(Some(event)) => info!("chain syncer received event: {:?}", event),
+            Poll::Pending | Poll::Ready(None) => (),
+        };
+        Poll::Pending
+    }
+}
+
+/// Message data used to ensure valid state transition
+struct MsgMetaData {
+    balance: BigUint,
+    sequence: u64,
+}
+
+impl<'db, DB> ChainSyncer<'db, DB, HamtStateTree>
+where
+    DB: BlockStore,
+    // BS: BlockSyncProvider,
```

Comment:
Can this be removed?

Reply:
ya will do, had kept in because I wanted to make the provider generic but probably not needed anymore with just channels passed in to handle requests